pkg/eval/ievaluator.go

```diff
@@ -1,15 +1,19 @@
 package eval
 
+import gen "github.com/open-feature/flagd/pkg/generated"
+
 /*
 IEvaluator implementations store the state of the flags,
 do parsing and validation of the flag state and evaluate flags in response to handlers.
 */
 type IEvaluator interface {
-	// TODO: add context param when rule evaluator is implemented
 	GetState() (string, error)
 	SetState(state string) error
-	ResolveBooleanValue(flagKey string, defaultValue bool) (bool, string, error)
-	ResolveStringValue(flagKey string, defaultValue string) (string, string, error)
-	ResolveNumberValue(flagKey string, defaultValue float32) (float32, string, error)
-	ResolveObjectValue(flagKey string, defaultValue map[string]interface{}) (map[string]interface{}, string, error)
+	ResolveBooleanValue(flagKey string, defaultValue bool, context gen.Context) (value bool, reason string, err error)
```

Comment:
More of a general observation, but the error message can contain a reason inside of it, what would the value be of having a separate error to a reason?

Reply:
There's a semantic difference between a reason and an error within the OpenFeature spec. The reason denotes why a flag evaluated the way it did, even in cases of normal execution. For example, a reason could denote that specific user targeting occurred, or a random split occurred, or a static value was used, etc. In abnormal execution, the reason should denote an error occurred, but the details of that error are carried by the error code itself.

https://github.com/open-feature/spec/blob/main/specification/flag-evaluation.md#requirement-146